src/speech_node/speech.py

```diff
@@ -26,54 +26,133 @@ class bcolors:
+        """
+
+        # Change speak topic message to same type as service call
         req = SpeakRequest()
         req.sentence = sentence_msg.data
-        # req.character = self.character
-        # req.language = self.language
-        # req.voice = self.voice
-        # req.emotion = self.emotion
-        req.blocking_call = True
 
-        self.do_tts(req)
+        # req.blocking_call = False  # False by default
+
+        self.buffer_requests(req)
 
     def speak_srv(self, req):
-        self.do_tts(req)
+        """
+        Receiving service calls over the ~speak service
+        :param req: text_to_speech.srv server message - SpeakRequest
+        """
+
+        self.buffer_requests(req)
         return ""
 
+    def clear_buffer_srv(self, empty):
+        """
+        Clearing the buffer on service call and setting the (now empty) queue to non-blocking to avoid dead-lock
+        :param empty: empty service call
+        """
+
+        self.buffer.clear()
+        self.block_queue = False
+        return []
+
+    def spin(self):
+        """
+        ROS spin-like function, should be ran after initializing the node. Sends new TTS goals to the Toyota TTS if
+        the buffer queue is non-empty and the robot isn't already talking and adjusts the buffer as necessary
+        """
+
+        while not rospy.is_shutdown():
+
+            # Print the talking state of the robot (PENDING = 0, TALKING = 1, DONE = 2) to the ROS debugging level
+            rospy.logdebug(self.speech_client.simple_state)
+
+            # If the buffer is non-empty and the robot is not currently talking, send a new TTS request
+            if self.buffer and not self.speech_client.simple_state == actionlib.SimpleGoalState.ACTIVE:
+                send_req = self.buffer[0]
+
+                # Bridge between the req object and simple_action_client goal object
+                goal = TalkRequestGoal()
+                out = Voice()
+                out.interrupting = False
+                out.queueing = True
+                out.language = 1
+                out.sentence = send_req.sentence
+                goal.data = out
+
+                # Send the left-most queue entry to Toyota TTS over the simple_action_client
+                self.speech_client.send_goal(goal)
+
+                # Pop left-most queue entry
+                if self.buffer:
+                    self.buffer.popleft()
+                rospy.logdebug("Buffer size [{}]: Send TTS request and removed from queue.".format(len(self.buffer)))
+
+                # If the last call is blocking then wait before setting block_queue to false, i.e. blocking until done.
+                # Might need to check code below for better alternatives to while-loop.
+                if send_req.blocking_call:
+                    while not rospy.is_shutdown() and not self.buffer and \
+                            self.speech_client.simple_state == actionlib.SimpleGoalState.ACTIVE:
```

Comment:
This does not seem like the 'simplest' implementation. A braindump:
* Remember the 'active' goal:
```python
self._active_req = self.buffer.popleft()
```
* Send the goal with a done callback that sets it back to None:
```python
def _done_cb(self, _, __):
    self._active_req = None
...
self.speech_client.send_goal(self._active_req, done_cb=self._done_cb)
```
* Turn 'block queue' into a property:
```python
@property
def block_queue(self):
    # First check if self._active_req is not None and if not, if it is blocking
    active_req_blocking = self._active_req is not None and self._active_req.blocking
    return active_req_blocking or any(requests.blocking_call for requests in self.buffer)
```
* With the done callback, you can change the main while loop to:
```python
while not rospy.is_shutdown():  # Loop until rospy is shutdown
    if self._active_req is not None or not self.buffer: 
        rate.sleep()
        continue

    # Start all the magic here
```